File: hdf-converters/converters/hdf-to-xccdf/go/groupid_test.go
package hdftoxccdf

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type groupIDCase struct {
	GID         string `json:"gid"`
	ID          string `json:"id"`
	Passthrough bool   `json:"passthrough"`
	Why         string `json:"why"`
}

func loadGroupIDCases(t *testing.T) []groupIDCase {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "shared", "xccdf-group-id-cases.json"))
	require.NoError(t, err)

	var table struct {
		Cases []groupIDCase `json:"cases"`
	}
	require.NoError(t, json.Unmarshal(raw, &table))
	require.NotEmpty(t, table.Cases, "an empty table would pass vacuously")
	return table.Cases
}

// The encoder is implemented twice, so the expectations live in one shared file
// both languages read rather than in two hand-kept copies.
func TestXCCDFGroupIDMatchesSharedTable(t *testing.T) {
	for _, c := range loadGroupIDCases(t) {
		t.Run(c.GID, func(t *testing.T) {
			assert.Equal(t, c.ID, xccdfGroupID(c.GID), c.Why)
			assert.Equal(t, c.Passthrough, isXCCDFGroupID(c.GID), c.Why)
		})
	}
}

// Whatever the input, the result must satisfy groupIdType — the property the
// encoder exists to guarantee, asserted here against the XSD's own pattern
// rather than against the encoder's idea of it.
func TestXCCDFGroupIDAlwaysSatisfiesGroupIDType(t *testing.T) {
	pattern := regexp.MustCompile(`^xccdf_[^_]+_group_.+$`)
	ncname := regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9._-]*$`)

	gids := []string{"", " ", "_", "café", "日本", "V-1", "a:b", "xccdf__group_x"}
	for _, c := range loadGroupIDCases(t) {
		gids = append(gids, c.GID)
	}
	for _, gid := range gids {
		id := xccdfGroupID(gid)
		assert.Regexp(t, pattern, id, "gid %q encoded to %q, which fails the XSD pattern", gid, id)
		assert.Regexp(t, ncname, id, "gid %q encoded to %q, which is not an NCName", gid, id)
	}
}

type realShape struct {
	GID    string `json:"gid"`
	Shape  string `json:"shape"`
	Source string `json:"source"`
}

func loadRealShapes(t *testing.T) []realShape {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "shared", "xccdf-group-id-cases.json"))
	require.NoError(t, err)

	var table struct {
		RealShapes []realShape `json:"realShapes"`
	}
	require.NoError(t, json.Unmarshal(raw, &table))
	require.NotEmpty(t, table.RealShapes, "an empty table would pass vacuously")
	return table.RealShapes
}

// Encoding is not injective: sanitization folds every character outside
// [A-Za-z0-9._-] to "_", so two gids that differ only in such characters
// collide, and the shared table above pins that behaviour on hostile input.
// What real content needs is narrower and provable rather than sampled: for
// every gid shape real tools emit, sanitization must be the identity, because
// then encoding is prefix-plus-gid (or passthrough) and cannot collide at all.
// One representative per shape is enough — the property is about the
// characters a shape uses, not about how many gids share it.
func TestXCCDFGroupIDLeavesRealShapesUntouched(t *testing.T) {
	for _, r := range loadRealShapes(t) {
		t.Run(r.Shape, func(t *testing.T) {
			got := xccdfGroupID(r.GID)
			if isXCCDFGroupID(r.GID) {
				assert.Equal(t, r.GID, got, "a conforming %s id must pass through unchanged", r.Source)
				return
			}
			assert.Equal(t, "xccdf_hdf_group_"+r.GID, got,
				"a %s id (%s) was rewritten by the sanitizer — that shape can now collide", r.Source, r.Shape)
		})
	}
}

File: hdf-converters/converters/hdf-to-xccdf/typescript/groupid.test.ts
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, it, expect } from 'vitest';
import { convertHdfToXccdf } from './converter.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

interface GroupIdCase {
  gid: string;
  id: string;
  passthrough: boolean;
  why: string;
}

const CASES = (
  JSON.parse(
    readFileSync(join(__dirname, '..', '..', '..', 'shared', 'xccdf-group-id-cases.json'), 'utf-8'),
  ) as { cases: GroupIdCase[] }
).cases;

function hdfWithGid(gid: string): string {
  return JSON.stringify({
    timestamp: '2020-01-01T00:00:00Z',
    baselines: [
      {
        name: 'b',
        requirements: [
          {
            id: 'r',
            impact: 0,
            tags: { gid },
            descriptions: [{ label: 'default', data: 'd' }],
            results: [{ status: 'passed', codeDesc: 'c', startTime: '2020-01-01T00:00:00Z' }],
          },
        ],
      },
    ],
  });
}

// The encoding is implemented twice, so the expectations live in one shared file
// both languages read. Asserted through the converter rather than against the
// helper directly, so this also pins that the encoder is actually wired up.
describe('hdf-to-xccdf Group/@id', () => {
  it('has a populated shared table', () => {
    expect(CASES.length, 'an empty table would pass vacuously').toBeGreaterThan(0);
  });

  it.each(CASES.filter((c) => c.gid !== '').map((c) => [c.gid, c] as const))(
    'encodes %j like the Go peer',
    (_gid, c) => {
      expect(convertHdfToXccdf(hdfWithGid(c.gid)), c.why).toContain(`id="${c.id}"`);
    },
  );

  // XCCDF types Group/@id as an NCName that must also match this pattern
  // (xccdf_1.2.xsd:821). Asserted against the XSD's own pattern rather than
  // against the encoder's idea of it; the Go peer gates on the real XSD.
  it.each(CASES.filter((c) => c.gid !== '').map((c) => [c.gid, c] as const))(
    'emits a groupIdType for %j',
    (_gid, c) => {
      const id = /<(?:cdf:)?Group[^>]*\bid="([^"]*)"/.exec(convertHdfToXccdf(hdfWithGid(c.gid)))?.[1];
      expect(id, 'no Group element was emitted').toBeDefined();
      expect(id).toMatch(/^xccdf_[^_]+_group_.+$/);
      expect(id).toMatch(/^[A-Za-z_][A-Za-z0-9._-]*$/);
    },
  );
});

// benchmarkIdType and profileIdType require a trailing name segment via their
// .+ (xccdf_1.2.xsd:799, :843). A baseline with an empty name is valid HDF but
// produced "xccdf_hdf_benchmark_", which the XSD rejects. The Go peer gates the
// same input on the real XSD.
describe('hdf-to-xccdf empty baseline name', () => {
  it('still emits ids with a trailing name segment', () => {
    const out = convertHdfToXccdf(
      JSON.stringify({
        timestamp: '2020-01-01T00:00:00Z',
        baselines: [
          {
            name: '',
            requirements: [
              {
                id: 'r',
                impact: 0,
                tags: {},
                descriptions: [{ label: 'default', data: 'd' }],
                results: [{ status: 'passed', codeDesc: 'c', startTime: '2020-01-01T00:00:00Z' }],
              },
            ],
          },
        ],
      }),
    );
    expect(out).toContain('id="xccdf_hdf_benchmark_unnamed"');
    expect(out).toContain('id="xccdf_hdf_profile_unnamed"');
    expect(out).toMatch(/id="xccdf_[^_"]+_benchmark_[^"]+"/);
  });
});

// Every timestamp this converter emits is HDF's canonical form: RFC3339 in UTC
// with trailing fractional zeros trimmed. This side used to pass the source
// string through, so a result time of ".500Z" emitted ".500Z" here and ".5Z" in
// Go — a divergence no fixture happened to carry. Asserts the exact strings the
// Go peer asserts.
describe('hdf-to-xccdf timestamps', () => {
  it('emits canonical trimmed-fraction times', () => {
    const out = convertHdfToXccdf(
      JSON.stringify({
        baselines: [
          {
            name: 'b',
            requirements: [
              {
                id: 'r',
                impact: 0,
                tags: {},
                descriptions: [{ label: 'default', data: 'd' }],
                results: [
                  { status: 'passed', codeDesc: 'c', startTime: '2020-01-02T09:00:00.500Z' },
                  { status: 'failed', codeDesc: 'c2', startTime: '2020-01-02T03:04:05.120Z' },
                  { status: 'passed', codeDesc: 'c3', startTime: '2020-01-02T07:30:00.000Z' },
                ],
              },
            ],
          },
        ],
      }),
    );

    expect(out).toContain('start-time="2020-01-02T03:04:05.12Z"');
    expect(out).toContain('end-time="2020-01-02T09:00:00.5Z"');
    expect(out).toContain('time="2020-01-02T07:30:00Z"');
    expect(out).not.toContain('.500Z');
    expect(out).not.toContain('.120Z');
    expect(out).not.toContain('.000Z');
  });
});

// The two languages guarded the InSpec <check> differently: Go emitted one
// whenever code was non-nil, so an empty string produced an empty check element,
// while this side's truthy test skipped it. Both now use the shared
// non-empty-text helper. Mirrors the Go peer case for case.
describe('hdf-to-xccdf InSpec check', () => {
  const INSPEC = 'system="http://inspec.io/"';

  it.each([
    ['absent', undefined, false],
    ['empty', '', false],
    ['whitespace only', '   ', false],
    ['real code', "describe file('/etc/passwd') do\n end", true],
  ])('emits a check for %s: %s', (_label, code, want) => {
    const out = convertHdfToXccdf(
      JSON.stringify({
        baselines: [
          {
            name: 'b',
            requirements: [
              {
                id: 'r',
                impact: 0,
                tags: {},
                ...(code === undefined ? {} : { code }),
                descriptions: [{ label: 'default', data: 'd' }],
                results: [{ status: 'passed', codeDesc: 'c', startTime: '2020-01-01T00:00:00Z' }],
              },
            ],
          },
        ],
      }),
    );
    expect(out.includes(INSPEC), 'a check is emitted only for code that carries text').toBe(want);
  });
});
